File: protocol/src/test/typescript/zfoots/ProtocolManager.ts
import EmptyObject from './packet/EmptyObject';
import { EmptyObjectRegistration } from './packet/EmptyObject';
import VeryBigObject from './packet/VeryBigObject';
import { VeryBigObjectRegistration } from './packet/VeryBigObject';
import ComplexObject from './packet/ComplexObject';
import { ComplexObjectRegistration } from './packet/ComplexObject';
import NormalObject from './packet/NormalObject';
import { NormalObjectRegistration } from './packet/NormalObject';
import ObjectA from './packet/ObjectA';
import { ObjectARegistration } from './packet/ObjectA';
import ObjectB from './packet/ObjectB';
import { ObjectBRegistration } from './packet/ObjectB';
import SimpleObject from './packet/SimpleObject';
import { SimpleObjectRegistration } from './packet/SimpleObject';
import IByteBuffer from "./IByteBuffer";
import IProtocolRegistration from "./IProtocolRegistration";

const protocols = new Map<number, IProtocolRegistration<unknown>>();
const protocolIdMap = new Map<any, number>();

// initProtocol
protocols.set(0, new EmptyObjectRegistration());
protocolIdMap.set(EmptyObject, 0);
protocols.set(1, new VeryBigObjectRegistration());
protocolIdMap.set(VeryBigObject, 1);
protocols.set(100, new ComplexObjectRegistration());
protocolIdMap.set(ComplexObject, 100);
protocols.set(101, new NormalObjectRegistration());
protocolIdMap.set(NormalObject, 101);
protocols.set(102, new ObjectARegistration());
protocolIdMap.set(ObjectA, 102);
protocols.set(103, new ObjectBRegistration());
protocolIdMap.set(ObjectB, 103);
protocols.set(104, new SimpleObjectRegistration());
protocolIdMap.set(SimpleObject, 104);

class ProtocolManager {
    static getProtocolId(clazz: any): number {
        const protocolId = protocolIdMap.get(clazz);
        if (protocolId === null || protocolId === undefined) {
            throw '[protocol:' + clazz + '] not exist';
        }
        return protocolId;
    }

    static getProtocol(protocolId: number): IProtocolRegistration<unknown> {
        const protocol = protocols.get(protocolId);
        if (protocol === null || protocol === undefined) {
            throw '[protocolId:' + protocolId + '] not exist';
        }
        return protocol;
    }

    static write(buffer: IByteBuffer, packet: any): void {
        const protocolId = ProtocolManager.getProtocolId(packet.constructor);
        buffer.writeShort(protocolId);
        const protocol = ProtocolManager.getProtocol(protocolId);
        protocol.write(buffer, packet);
    }

    static read(buffer: IByteBuffer): any {
        const protocolId = buffer.readShort();
        const protocol = ProtocolManager.getProtocol(protocolId);
        const packet = protocol.read(buffer);
        return packet;
    }
}

export default ProtocolManager;

File: protocol/src/test/dart/zfoodart/ProtocolManager.dart
import './IProtocolRegistration.dart';
import 'IByteBuffer.dart';
import 'packet/EmptyObject.dart';
import 'packet/VeryBigObject.dart';
import 'packet/ComplexObject.dart';
import 'packet/NormalObject.dart';
import 'packet/ObjectA.dart';
import 'packet/ObjectB.dart';
import 'packet/SimpleObject.dart';


class Protocolmanager {
  static Map<int, IProtocolRegistration> protocols = Map();
  static Map<Object, int> protocolIdMap = Map();


  static void initProtocol() {
    // initProtocol
    protocols[0] = EmptyObjectRegistration();
    protocolIdMap[EmptyObject] = 0;
    protocols[1] = VeryBigObjectRegistration();
    protocolIdMap[VeryBigObject] = 1;
    protocols[100] = ComplexObjectRegistration();
    protocolIdMap[ComplexObject] = 100;
    protocols[101] = NormalObjectRegistration();
    protocolIdMap[NormalObject] = 101;
    protocols[102] = ObjectARegistration();
    protocolIdMap[ObjectA] = 102;
    protocols[103] = ObjectBRegistration();
    protocolIdMap[ObjectB] = 103;
    protocols[104] = SimpleObjectRegistration();
    protocolIdMap[SimpleObject] = 104;
  }

  static int getProtocolId(Object clazz) {
    var protocolId = protocolIdMap[clazz];
    if (protocolId == null) {
      throw Exception("protocol:[$protocolId] not exist");
    }
    return protocolId;
  }

  static IProtocolRegistration getProtocol(int protocolId) {
    var protocol = protocols[protocolId];
    if (protocol == null) {
      throw Exception("protocol:[$protocolId] not exist");
    }
    return protocol;
  }

  static void write(IByteBuffer buffer, Object packet) {
    var protocolId = getProtocolId(packet.runtimeType);
    buffer.writeShort(protocolId);
    var protocol = getProtocol(protocolId);
    protocol.write(buffer, packet);
  }

  static Object read(IByteBuffer buffer) {
    var protocolId = buffer.readShort();
    var protocol = getProtocol(protocolId);
    var packet = protocol.read(buffer);
    return packet;
  }
}